Add common issue docs for PEP585 syntax update

LefterisJP opened the pull request:
As discussed with @AlexWaygood in https://github.com/python/mypy/issues/14245#issuecomment-1336397899 adding a common issue entry for the problem of hitting overshadowing a built-in keyword when upgrading to PEP584 syntax

LefterisJP committed:
Add common issue docs for PEP585 syntax upgrade

As discussed with @AlexWaygood in
https://github.com/python/mypy/issues/14245#issuecomment-1336397899
adding a common issue entry for the problem of hitting overshadowing a
built-in keyword when upgrading to PEP585 syntax

AlexWaygood commented on docs/source/common_issues.rst:
```diff
@@ -821,3 +821,32 @@ This is best understood via an example:
+
+In pre-PEP584 syntax type annotations for builtin types such as ``dict``, ``list``, ``tuple``, ``type`` were taken from the ``typing`` module as ``typing.Dict``, ``typing.List``, ``typing.Tuple``, ``typing.Type`` etc.
+
+When upgrading a codebase you may hit a case such as the following:
```
```suggestion
When updating a codebase to use modern syntax, you may hit a case
such as the following:
```

AlexWaygood commented on docs/source/common_issues.rst:
```diff
@@ -821,3 +821,32 @@ This is best understood via an example:
+
+    from typing import TypeVar
+
+    T = TypeVar('T', bound='Foo')
```
Mypy uses `black` style across its codebase; let's stick to that here

```suggestion
    T = TypeVar("T", bound="Foo")
```

AlexWaygood commented on docs/source/common_issues.rst:
```diff
@@ -821,3 +821,32 @@ This is best understood via an example:
+	type: str
+
+	@classmethod
+	def foo(cls: type[T], value: str) -> None:  # will fail due to "not a valid type"
```
- It's somewhat confusing for the method to be called `foo()` when the class is called `Foo`.
- Be slightly more specific about mypy's complaint here

```suggestion
	def bar(cls: type[T], value: str) -> None:  # Mypy complains that "Foo.type" is "not valid as a type"
```

LefterisJP replied:
I moved the comment here to the pass statement to not make it above 80 chars and have the sliding bar appear on the bottom of the code block. I am not sure if this is preferred. Let me know.

AlexWaygood commented on docs/source/common_issues.rst:
```diff
@@ -821,3 +821,32 @@ This is best understood via an example:
+	def foo(cls: type[T], value: str) -> None:  # will fail due to "not a valid type"
+	    pass
+
+This is happening since the class has an attribute that is overshadowing the built-in keyword ``type``. To get this code to type-check you have the following options:
```
- `type` is a builtin name, but it's not a _keyword_ according to the language specification. (I _think_ this is the correct intersphinx reference to link to the CPython docs for `type`; apologies if not.)
- I think a slightly longer example in the interactive REPL might be helpful here to illustrate exactly what's going on with the scoping semantics.

```suggestion
This occurs due to the fact that the class has an attribute named `type`
that shadows the built-in function :py:func:`type`.
To illustrate: at runtime in the following example,
the annotation for `arg` in the `ham` method 
is resolved as being the string `"spam"`
due to the fact that method signatures are executed
in the same scope as class attributes:

.. code-block:: pycon

    >>> class Eggs:
    ...     type: str = "spam"
    ...     def ham(self, arg: type) -> None:
    ...         pass
    ...
    >>> import inspect
    >>> inspect.signature(Eggs.ham)
    <Signature (self, arg: 'spam') -> None>

Mypy follows the same scoping semantics as Python at runtime
when resolving references.
To get this code to type-check, therefore, you have the following options:
```

LefterisJP replied:
> type is a builtin name, but it's not a keyword according to the language specification. (I think this is the correct intersphinx reference to link to the CPython docs for type; apologies if not.)

True. Thanks will change it. The issue I have with the link is that it does not work/render when creating the html docs. I tried to grep mypy docs for usage of type and I already see the same problem at https://mypy.readthedocs.io/en/stable/type_narrowing.html. The `type()` link is the only one not working. It's using the same exact schema you gave here.

>  think a slightly longer example in the interactive REPL might be helpful here to illustrate exactly what's going on with the scoping semantics.

Very nice idea. I agree and done it. It also helped me better understand it :)

One question, what does `pycon` stand for in this code block?

AlexWaygood replied:
> One question, what does `pycon` stand for in this code block?

"Python console" -- it provides specialised syntax highlighting for the interactive REPL :)

AlexWaygood commented on docs/source/common_issues.rst:
```diff
@@ -821,3 +821,32 @@ This is best understood via an example:
+
+This is happening since the class has an attribute that is overshadowing the built-in keyword ``type``. To get this code to type-check you have the following options:
+
+- Use ``typing.Type`` and use ``Type[]`` for your annotation.
```
```suggestion
- Continue using `typing.Type[]` for your annotation.
```

AlexWaygood commented on docs/source/common_issues.rst:
```diff
@@ -821,3 +821,32 @@ This is best understood via an example:
+This is happening since the class has an attribute that is overshadowing the built-in keyword ``type``. To get this code to type-check you have the following options:
+
+- Use ``typing.Type`` and use ``Type[]`` for your annotation.
+- Stick ``import builtins`` at the top and use ``builtins.type[]`` for your annotation
```
My wording in the issue was slightly informal; that's fine for an issue, but we can do better in docs :)

```suggestion
- Add ``import builtins`` to the top of your module and use ``builtins.type[]`` for your annotation
```

AlexWaygood commented on docs/source/common_issues.rst:
```diff
@@ -821,3 +821,32 @@ This is best understood via an example:
+
+- Use ``typing.Type`` and use ``Type[]`` for your annotation.
+- Stick ``import builtins`` at the top and use ``builtins.type[]`` for your annotation
+- Add ``from builtins import type as Type`` at the top and use ``Type[]`` for your annotation
```
```suggestion
- Add ``from builtins import type as Type`` to the top of your module and use ``Type[]`` for your annotation
```

AlexWaygood commented on docs/source/common_issues.rst:
```diff
@@ -821,3 +821,32 @@ This is best understood via an example:
+- Stick ``import builtins`` at the top and use ``builtins.type[]`` for your annotation
+- Add ``from builtins import type as Type`` at the top and use ``Type[]`` for your annotation
+
+The same thing applies for any other built-in keyword.
```
```suggestion
The same thing applies for any other built-in name.
```

AlexWaygood reviewed (commented):
Thanks! I've left a few suggestions below.

Main points:

- It's PEP 585, not PEP 584 :)
- Stick to <=80 characters per line where possible.
- `type` is a builtin name, but it's not a _keyword_
- Use Sphinx roles to add links where appropriate

LefterisJP commented:
> Thanks! I've left a few suggestions below.
> 
> Main points:
> 
> * It's PEP 585, not PEP 584 :)
> * Stick to <=80 characters per line where possible.
> * `type` is a builtin name, but it's not a _keyword_
> * Use Sphinx roles to add links where appropriate

Thanks I addressed all your points. Didn't know of the sphinx roles and links. The off by one error is funny :rofl: 

I left you two extra questions

AlexWaygood commented:
> The off by one error is funny 🤣

You need to update the title of the PR as well ;)

AlexWaygood commented on docs/source/common_issues.rst:
```diff
@@ -821,3 +821,55 @@ This is best understood via an example:
+Upgrading to :pep:`585` syntax
+------------------------------
+
+In pre-:pep:`585` syntax type annotations for builtin types such as ``dict``,
+``list``, ``tuple``, ``type`` were taken from the :py:mod:`typing` module as
+``typing.Dict``, ``typing.List``, ``typing.Tuple``, ``typing.Type`` etc.
```
Minor wordsmithing:

```suggestion
Prior to :pep:`585`, instances of builtin types such as ``dict``, ``list``,
``tuple``, and ``type`` were annotated using aliases
from the :py:mod:`typing` module
(``typing.Dict``, ``typing.List``, ``typing.Tuple``, ``typing.Type`` etc.).
```

AlexWaygood commented on docs/source/common_issues.rst:
```diff
@@ -821,3 +821,55 @@ This is best understood via an example:
+    class Foo:
+	type: str
+
+	@classmethod
+	def bar(cls: type[T], value: str) -> None:
+	    pass    # Mypy complains that "Foo.type" is "not valid as a type"
```
I'd put the comment above the method signature rather than below it.

```suggestion
        # Mypy complains that "Foo.type" is "not valid as a type"
	@classmethod
	def bar(cls: type[T], value: str) -> None:
	    pass
```

AlexWaygood commented on docs/source/common_issues.rst:
```diff
@@ -821,3 +821,55 @@ This is best understood via an example:
+To get this code to type-check, therefore, you have the following options:
+
+- Continue using  ``typing.Type[]`` for your annotation.
+- Add ``import builtins`` to the top of your module and use ``builtins.type[]`` for your annotation.
+- Add ``from builtins import type as Type`` to the top of your module and use ``Type[]`` for your annotation
```
<=80 chars :)

```suggestion
- Add ``import builtins`` to the top of your module and use ``builtins.type[]``
  for your annotation.
- Add ``from builtins import type as Type`` to the top of your module
  and use ``Type[]`` for your annotation
```

AlexWaygood reviewed (commented):
Thanks! A few more points:

LefterisJP commented:
> You need to update the title of the PR as well ;)

Done!

> Thanks! A few more points:

Addressed them and pushed.

One question from my side. [Here](https://github.com/python/mypy/pull/14246#discussion_r1038989681) I asked you if you know the sphinx link to type for the docs. Any idea? This does not seem to render properly. But I see the same problem in other places in the docs as per my comment.

AlexWaygood commented:
> One question from my side. [Here](https://github.com/python/mypy/pull/14246#discussion_r1038989681) I asked you if you know the sphinx link to type for the docs. Any idea? This does not seem to render properly. But I see the same problem in other places in the docs as per my comment.

Hmm, maybe try
````
py:class:`type`
````

LefterisJP commented:
> > One question from my side. [Here](https://github.com/python/mypy/pull/14246#discussion_r1038989681) I asked you if you know the sphinx link to type for the docs. Any idea? This does not seem to render properly. But I see the same problem in other places in the docs as per my comment.
> 
> Hmm, maybe try
> 
> ```
> :py:class:`type`
> ```

Yeah! That seems to work! Generates a proper link to: https://docs.python.org/3/library/functions.html#type

Follow up question then. There is one more place in the docs where this needs fixing. Should I do it in this PR and a separate commit? A different PR? I am not sure what workflow you guys prefer.

AlexWaygood commented:
> Follow up question then. There is one more place in the docs where this needs fixing. Should I do it in this PR and a separate commit? A different PR? I am not sure what workflow you guys prefer.

That would be a very welcome fix, but definitely best to do it as a separate PR, since it's a different issue to the thing being tackled here :)

AlexWaygood commented on docs/source/common_issues.rst:
```diff
@@ -821,3 +821,59 @@ This is best understood via an example:
 function.
+
+
+Upgrading to :pep:`585` syntax
+------------------------------
```
This is slightly more verbose, but I think it's good to have "name shadowing" in the section title, since that's the root cause of the issue:

```suggestion
Name shadowing issues with :pep:`585` syntax
--------------------------------------------
```

AlexWaygood commented on docs/source/common_issues.rst:
```diff
@@ -821,3 +821,59 @@ This is best understood via an example:
+Upgrading to :pep:`585` syntax
+------------------------------
+
+Prior to :pep:`585`, instances of builtin types such as ``dict``, ``list``,
```
The mypy docs generally seem to use "built-in" over "builtin" when it's used as an adjective

```suggestion
Prior to :pep:`585`, instances of built-in types such as ``dict``, ``list``,
```

AlexWaygood reviewed (commented):
Looks great now! Just two minor points:

LefterisJP commented:
> Looks great now! Just two minor points:

Addressed both. And also opened that follow-up PR.

AlexWaygood reviewed (approved):
LGTM!

LefterisJP commented:
> It's a bit weird to frame this around PEP 585, when PEP 585 doesn't really do anything special here: you can run into the exact same problem with types like `int` and `str` that have always been usable as annotations. Indeed there is already a section in "Common issues" about this: https://mypy.readthedocs.io/en/stable/common_issues.html#dealing-with-conflicting-names. So if anything, we should enhance that section.
> 
> I would also recommend removing the recommendation to use `typing.Type`. One of the effects of PEP 585 is that things like "type" and "list" are no longer special in the sense that they have a separate typing alter ego. That's good for learning because users no longer have to remember which names are special like that. So we shouldn't point users to `typing.Type`; people working in a post-PEP 585 world shouldn't need to know it even exists.

Thank you for the review @JelleZijlstra @AlexWaygood. So how would you like me to proceed with the PR to get it merged? Add the section as is under the section you recommended removing:

1. references to PEP584?
2. The recommendation to use `typing.Type` ?

JelleZijlstra commented:
Yes, I'd recommend adding a shortened text under the existing section. It's probably helpful to use `type` as the main example since that's in my experience the most common name for which this problem comes up.

hauntsaninja commented:
This is stale and not sure it is needed
